[Console/Python 3] Remove pyreadline/gnureadline

gpotter2 opened the pull request:
This PR is an attempt of removing pyreadline/gnureadline, which have not been updated since (at least) 2015, and are mostly broken under python 3...
- Removes pyreadline/gnureadline. Now use `IPython` to handle the shell.
![image](https://user-images.githubusercontent.com/10530980/30378245-dd4cefb8-9892-11e7-9ca4-37b83f68bb22.png)

gpotter2 commented:
@ThomasFaivre Could you have a look at this please ? Thanks!

gpotter2 commented on scapy/utils.py:
```diff
@@ -37,6 +37,9 @@ def get_temp_file(keep=False, autoext=""):
     return f
 
+class Bunch:
+    __init__ = lambda self, **kw: setattr(self, '__dict__', kw)
```
This may be used in tests:
```
>>> Bunch(a=1).a
1
```

p-l- commented:
That looks promising! Can you rebase against master?

gpotter2 commented:
Recases.

p-l- commented:
I have a question for you and @guedou: is it still useful, now, to have this rather than using dedicated shells, like iPython for example. What do you think?

gpotter2 commented:
I guess IPython could handle it...

We simply need to have a backup when it's not available...

guedou commented:
I don't have a strong opinion on this topic. As long as the module is easy to install and use on our side, I am fine with any choice. 
If most of the same features can be achieved with iPython without any modifications in Scapy, that might be the best option.

p-l- commented:
If people don't like iPython, they have at least bpython as an alternative. I don't think we really need to maintain this interface when great tools exist that do the job (and can run under Linux, *BSD, MacOS & Windows).

@gpotter2 are you one of the people who "don't like iPython"? I'd like to hear what you don't like if that's the case. Thanks!

gpotter2 commented:
@p-l- I don't like the input system "In[..]" "Out...". I saw there is an option to change this to the "classic legacy" prompt, so i'm ok to use IPython instead.

p-l- commented:
@gpotter2 maybe we could have `run_scapy{,.bat}` that runs iPython with modified prompt to mimic current behavior?

gpotter2 commented:
@p-l- Sounds good. Will work onto this.

gpotter2 commented:
@p-l- I've updated the shell.
Could you please try it and give your feedback ? Thanks !

gpotter2 commented:
Fixed the bug "No warning messages were shown", added indirectedly by @p-l- in https://github.com/secdev/scapy/pull/764/files.
I added a warning in the logs so this will be known in the future :)

guedou commented:
There are some changes to the current behavior:
- the prompt ">>>" is not white anymore
- the shell asks "Do you really want to exit ([y]/n)?"
- the history is not written to ~/.scapy_history

I think that they should be fixed.

gpotter2 commented:
@guedou Fixed all of your comments !
If you are willing to change the colors, tell me. I do find the white a bit weird...

p-l- commented:
Great job! One problem I have with this code for now: when I type: `IP()/[tab]`, iPython tries to complete with files at the root of my file system (`/bin/`, etc.). I have to type `IP() / [tab]` (note the spaces), it works, so `IP() / IC[tab]` will complete in `ICMP`.

gpotter2 commented:
Well I don't know how to fix this :/
I'll ask on the Ipython github page i guess...

p-l- commented:
This patch will add the fields as a completion on packet objects. The clean way to do that is normally to create a `Packet.__dir__(self)` method, but I don't know how to add the fields to the "normal" result of `dir()`.

```patch
diff --git a/scapy/main.py b/scapy/main.py
index 38d37ac..23c0d83 100644
--- a/scapy/main.py
+++ b/scapy/main.py
@@ -314,9 +314,15 @@ def interact(mydict=None,argv=None,mybanner=None,loglevel=20):
             IPYTHON=False
         
     if IPYTHON:
+        from IPython.terminal.embed import InteractiveShellEmbed
         from IPython.terminal.prompts import Prompts, Token
+        from IPython.utils.generics import complete_object
         from traitlets.config.loader import Config
-        from IPython.terminal.embed import InteractiveShellEmbed
+        from scapy.packet import Packet
+
+        @complete_object.when_type(Packet)
+        def complete_packet(obj, prev_completions):
+            return prev_completions + [fld.name for fld in obj.fields_desc]
 
         banner = the_banner % (conf.version) + " using IPython %s" % IPython.__version__
         cfg = Config()
```

gpotter2 commented:
Added a pickle test (used to be performed by pyreadline)

p-l- commented:
@gpotter2 if you do ask questions on Ipython's Github, could you link to this issue so that we can track the answers? Thanks!

gpotter2 commented:
@p-l- All those issues are related to the LTS IPython for python 2.7 only branch. They have all been fixed on the latest python 3 only branch.

I'm not sure they will add it those fixes to the LTS, I asked them to do so.

If not, I'll try to monkey-patch IPython autocompletion for Python 2.7

gpotter2 commented:
@p-l- @guedou Updated according to the new logo PR !
Also, colors are now available (on windows) without any requirement + conf.color... is now used on IPython

guedou commented on scapy/autorun.py:
```diff
@@ -58,7 +58,7 @@ def autorun_commands(cmds, my_globals=None, ignore_globals=None, verb=0):
                     sys.stderr.write(sys.__dict__.get("ps2","... "))
                 else:
-                    sys.stderr.write(str(sys.__dict__.get("ps1",ColorPrompt())))
+                    sys.stderr.write(str(sys.__dict__.get("ps1",sys.ps1)))
```
Please put a space after the comma.

guedou commented on scapy/config.py:
```diff
@@ -326,26 +326,20 @@ def isCryptographyAdvanced():
-
-def _prompt_changer(attr,val):
-    prompt = conf.prompt
+def _prompt_changer(attr, val):
```
Could you add the following docstring:
""Change the current prompt theme"""

gpotter2 replied:
Done

guedou commented on scapy/data.py:
```diff
@@ -188,7 +188,7 @@ def load_manuf(filename):
     TCP_SERVICES,UDP_SERVICES=load_services(os.environ["SystemRoot"] + "\system32\drivers\etc\services")
     # Default value, will be updated by arch.windows
-    MANUFDB = load_manuf(os.environ["ProgramFiles"] + "\\wireshark\\manuf")
+    MANUFDB = None
```
Why are you modifying this line ? This should be in a dedicated PR if really needed.

guedou commented on scapy/main.py:
```diff
@@ -10,27 +10,23 @@
+    pass
 
-from scapy.config import conf
+# Do not add any imports here that would trigger a warning messsage
```
You mean: 'never ever add imports' or 'never add imports before this line' ? I think that you should make this textt more explicit.

gpotter2 replied:
# Never add global imports, in main.py, that would trigger a warning messsage
# Before the console handlers gets added in interact()

guedou commented on scapy/main.py:
```diff
@@ -58,13 +54,17 @@ def _probe_config_file(cf):
         return cf_path
 
-def _read_config_file(cf):
+def _read_config_file(cf, _globals=globals(), _locals=locals(), interactive=True):
```
Could you add a docstring and explain what the kw args do ?

gpotter2 replied:
Done

guedou commented on scapy/main.py:
```diff
@@ -182,25 +182,29 @@ def list_contrib(name=None):
 
 
 def save_session(fname=None, session=None, pickleProto=-1):
```
Could you add a docstring ?

For example, """save current Scapy session to the file specified in the fname arg"""

gpotter2 replied:
Done

guedou commented on scapy/main.py:
```diff
@@ -314,32 +321,19 @@ def _len(line):
     conf.color_theme = DefaultTheme()
+    conf.interactive = True
+    if loglevel is not None:
+        conf.logLevel=loglevel
```
Add space around =

guedou commented on scapy/themes.py:
```diff
@@ -268,16 +288,12 @@ class HTMLTheme2(HTMLTheme):
-                return ct.prompt(config.conf.prompt)
-        except:
-            return self.__prompt
+def apply_ipython_color(shell):
```
Docstring ? =)

gpotter2 replied:
Done

guedou commented on scapy/utils.py:
```diff
@@ -35,7 +35,7 @@ def get_temp_file(keep=False, autoext=""):
     if not keep:
         conf.temp_files.append(f+autoext)
-    return f
+    return f+autoext
```
Add spaces around +

guedou commented on test/pipetool.uts:
```diff
@@ -264,6 +264,8 @@ os.unlink("t2.pcap")
 = Test InjectSink and Inject3Sink
 ~ needs_root
 
+import mock
```
Is this needed in this PR ?

gpotter2 replied:
This is very small, A forgotten fix that provided from running the tests from the command line on windows. So small i thought i would just leave this in this PR.

gpotter2 commented:
- Removed trailing line:
- before
![image](https://user-images.githubusercontent.com/10530980/30378245-dd4cefb8-9892-11e7-9ca4-37b83f68bb22.png)
- now
![image](https://user-images.githubusercontent.com/10530980/30610029-4ba5a100-9d7d-11e7-98ab-55b66bbc9c0e.png)

@p-l- It would be great if this could be merged in a near future, as it needs to replace Pyreadline/GnuReadline on the Python 3 port.

gpotter2 commented on test/regression.uts:
```diff
@@ -1413,10 +1413,11 @@ a=Ether(str(Ether()/IPv6()/TCP()))
 
 = IPv6 Class binding with GRE - build
-str(IP()/GRE()/Ether()/IP()/GRE()/IPv6()) == b'E\x00\x00f\x00\x01\x00\x00@/|f\x7f\x00\x00\x01\x7f\x00\x00\x01\x00\x00eX\xff\xff\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x08\x00E\x00\x00@\x00\x01\x00\x00@/|\x8c\x7f\x00\x00\x01\x7f\x00\x00\x01\x00\x00\x86\xdd`\x00\x00\x00\x00\x00;@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00 ...
+s = str(IP(src="127.0.0.1")/GRE()/Ether(dst="ff:ff:ff:ff:ff:ff", src="00:00:00:00:00:00")/IP()/GRE()/IPv6(src="::1"))
```
In some cases, IPv6 resolution could get called through this function, which is not available on travis/appveyor

guedou commented:
I am fine with @gpotter2 changes. This PR can be merged as soon as the OS X tests stop.
